src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -55,22 +51,34 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
 
                 state.Current.CollectionPropertyInitialized = true;
             }
-            else
+            else if (state.Current.IsProcessingObject(ClassType.Object))
```

Comment:
It's because we could have pushed a dictionary or object property onto the ReadStack at the start of this method. We don't push only if we are looking at a root object.

This means we could be examining a different property/entity at the time of the second check.

BTW, we don't need the first `state.Current.IsProcessingObject(ClassType.Object)` at the top of this method, so I'm reverting this addition, and replacing it with a debug assert.
The reason we don't need it is:
- we just checked for ClassType.Enumerable in the `if` branch above.
- we know we aren't `ClassType.Value`, `ClassType.Unknown`, or `ClassType.Dictionary`:

https://github.com/dotnet/corefx/blob/4c3bd8a0533a2e35df371b91045a1a5f0ada7d54/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs#L63-L70

https://github.com/dotnet/corefx/blob/4c3bd8a0533a2e35df371b91045a1a5f0ada7d54/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs#L71-L74

- this leaves us with `ClassType.Object`.